pkg/tasks/update_template_content.go

```diff
@@ -329,3 +348,215 @@ func getRHRepoContentPath(rawURL string) (string, error) {
+		return nil, err
+	}
+	if env != nil {
+		return env, nil
+	}
+	env, err = t.cpClient.CreateEnvironment(envID, envID, prefix)
+	if err != nil {
+		return nil, err
+	}
+	return env, nil
+}
+
+func (t *UpdateTemplateContent) promoteContent(reposAdded []string, envID string) error {
+	var addedIDs []string
+	for _, repoUUID := range reposAdded {
+		addedIDs = append(addedIDs, candlepin_client.GetContentID(repoUUID))
+	}
+	err := t.cpClient.PromoteContentToEnvironment(envID, addedIDs)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (t *UpdateTemplateContent) demoteContent(reposRemoved []string, envID string) error {
+	var removedIDs []string
+	for _, repoUUID := range reposRemoved {
+		removedIDs = append(removedIDs, candlepin_client.GetContentID(repoUUID))
+	}
+	err := t.cpClient.DemoteContentFromEnvironment(envID, removedIDs)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// getContentList return the list of ContentDTO that will be created in Candlepin
+func (t *UpdateTemplateContent) getContentList() ([]caliri.ContentDTO, []string, error) {
+	uuids := strings.Join(t.payload.RepoConfigUUIDs, ",")
+	repos, _, err := t.daoReg.RepositoryConfig.List(t.orgId, api.PaginationData{Limit: -1}, api.FilterData{UUID: uuids, Origin: config.OriginRedHat + "," + config.OriginExternal})
+	if err != nil {
+		return nil, nil, err
+	}
+	return createContentItems(repos.Data, t.payload.TemplateUUID)
+}
+
+func (t *UpdateTemplateContent) updatePayload() error {
+	var err error
+	a := t.payload
+	t.task, err = (*t.queue).UpdatePayload(t.task, a)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func createContentItems(repos []api.RepositoryResponse, templateUUID string) ([]caliri.ContentDTO, []string, error) {
+	var content []caliri.ContentDTO
+	var contentIDs []string
+	for _, repo := range repos {
+		if repo.LastSnapshot == nil {
+			continue
+		}
+		id := candlepin_client.GetContentID(repo.UUID)
+		repoType := candlepin_client.YumRepoType
+		repoLabel, err := getRepoLabel(repo)
+		if err != nil {
+			return nil, nil, err
+		}
+		repoVendor := getRepoVendor(repo)
+		contentURL, _, err := getDistPathAndName(repo, templateUUID, "")
+		if err != nil {
+			return nil, nil, err
+		}
+
+		content = append(content, caliri.ContentDTO{
+			Id:         &id,
+			Type:       &repoType,
+			Label:      &repoLabel,
+			Name:       &repo.Name,
+			Vendor:     &repoVendor,
+			ContentUrl: &contentURL,
+		})
+		contentIDs = append(contentIDs, id)
+	}
+	return content, contentIDs, nil
+}
+
+func getRepoLabel(repo api.RepositoryResponse) (string, error) {
+	// Replace any nonalphanumeric characters with an underscore
+	// e.g: "!!my repo?test15()" => "__my_repo_test15__"
+	re, err := regexp.Compile(`[^a-zA-Z0-9:space]`)
+	if err != nil {
+		return "", err
+	}
+
+	var repoLabel string
+	if repo.OrgID == config.RedHatOrg {
+		repoLabel = repo.Label
+	} else {
+		repoLabel = re.ReplaceAllString(repo.Name, "_")
+	}
+	return repoLabel, nil
+}
```

Comment:
I think we'd need to handle the edge case where two different custom repos would map to the same label

Reply:
Is this case you're talking about? I don't get any errors creating things in this scenario, but I'm assuming this is an issue when consuming the config.repo file.

org1 with repo "my repo" => "my_repo"
org1 with repo "my?repo" => "my_repo"

I'm assuming for different orgs it's okay?

Reply:
Do you think it makes sense to check this in the task and return an error if there's conflicting labels? And then I guess later on we'd somehow expose that error after template creation.

Reply:
i'm kinda surprised candlepin doesn't prevent this! but yes, you wouldn't be able to consume both those on the client.  ideally this doesn't cause an error, but may require us to add some randomness to it if its already present.  

For example, 
"my repo"  ->  my_repo
"my?repo" -> my_repo_jrn3oiuf

its an edge case, so ideally not super common

Reply:
I pushed a change here, but I realize now this says to only add randomness _if already present_, so I need to go back and change that